sql/core/src/main/scala/org/apache/spark/sql/execution/Exchange.scala

```diff
@@ -208,6 +208,19 @@ private[sql] case class EnsureRequirements(sqlContext: SQLContext) extends Rule[
     }
   }
 
+  private def removeRedundantShuffle(childPlan: SparkPlan): SparkPlan = {
+    // For an query like df1.repartition(1000).join(df2, "_1"), we will repartition df1
+    // with shuffle. Then we perform shuffle again as part of the exchange added here.
+    // To avoid the extra shuffle, we should remove repartition operators if they are
+    // the child of Exchange and shuffle=True.
+    val optimized = childPlan match {
+      case Repartition(n, true, plan) => plan
+      case Exchange(p, plan) => plan
```

Comment:
Hmm, because `logical.RepartitionByExpression` will be transformed to `execution.Exchange` as well. I am wondering if it is possible to have an `Exchange` here.